adr/2025_03_10-suggestions/suggestions-as-history.md

```diff
@@ -0,0 +1,31 @@
+# Storing changes as separate _document histories_ and diffs between them
+
+Branched document histories is the idea that you can think of a suggestion of the document as having a "different history" of a document; where, a user made some changes to a copy of the document, which can later be merged back into the document somehow.
+
+![Branched document histories](./images/doc+suggestions.png)
+
+This would work very similarly to how the version diffing would work, and is intended to mostly re-use that logic.
+
+## Implementation
+
+To implement this, each user who suggests changes is given a separate document which they would have access to modify the contents of. This allows the user to own that document, while guaranteeing the original document stays unmodified.
+
+![Branched document histories merged](./images/doc+suggestions_merged.png)
+
+When viewing what suggestions are available on a document, the frontend would have to "collect" the suggestions from each suggestion document, apply an updates those documents may not have seen yet, diff the documents to see what changes are in the suggestion document compared to the original document and finally display those changes to the user in the original document somehow.
+
+## Pros
+
+- Allows reuse of the version comparison logic
+- A suggestion-only user is easy to implement, since they can be given access to a copy of the Ydoc, without affecting the original document
+- Guaranteed that a suggestion cannot affect the original document without any server-side inspection of the content
+
+## Cons
+
+- Requires diffing an entire YDoc to derive what the change actually was
+  - Performance might be an issue especially since it scales by number of users who suggest
+- Unclear whether to diff the content at Y.js level or to diff the JSON document
+  - If diffing at the Y.js level, not everyone uses Y.js and we may not want to pay that bundle/complexity cost
+  - If diffing on the JSON, are we just recreating the version snapshot comparison and duplicating the work involved?
+- Complicated to setup
+  - Storage of the branched versions need to be kept up to date (i.e. must apply new updates since the branch) and need to store that document separately
```

Comment:
Perhaps, we could have a hybrid approach, where suggestions are stored into a different document (as a different history), but when in Prosemirror, we lie to Prosemirror & represent them as marks, this would allow us to re-use the approaches around tracked changes at the Prosemirror level, keep all of the benefits of marks just naturallly being transformed, and defer two sources of truth to a separate layer like y-prosemirror to sync the suggestions document and the root document.

Reply:
Reason being that it is quite unclear how to have 2 sources of truth for editing a single Prosemirror document, who is responsible for synchronizing the change? The problem with the editor having to be aware of this is that it potentially complicates all of the editor content to have to become "suggestion-aware", like if you are editing in a paragraph it has to choose where to save that change into.